### ISSUE TITLE
User documentation - broad discussion of purpose, structure and method

### ISSUE THREAD

**Issue by gwyndaf** [ChordPro/chordpro#490]
# Big Picture Overview

Here are the three main areas into which I think we can organise user documentation. This thread is for discussing that broad overall thinking (not content of individual sections). 

For each of the three main areas below, I'll create a _separate_ discussion thread/issue. First of these will be an outline for the Tutorial section. It's likely that Github discussion will skew toward developer input (which is valuable for accuracy), so I'll also seek input in the User Forum.

The broad purpose, particularly for Tutorial and Cookbook, is to make basic Chord usage more accessible, and expand the user base to include more non-technical users, without making it harder for experienced users who want more power and control.

## 1. Tutorial

**Rationale:** Expand the ChordPro user base, by making basic usage and concepts easier to understand for non-technical users. Prospective users could easily be deterred by technical concepts and methods, so the overall aim of the tutorial is provide a clear and non-technical introduction to understanding and using ChordPro, both the format and the software. It's likely this section will be read only once or twice, at first use.

**Focus:** User-centred, *i.e.* building users' conceptual understanding and practical skills.

**Target audience:** Users with a musical but non-technical background, familiar with GUI software on Windows or MacOS (let's assume Linux users are pretty tech-savvy), wanting to produce PDF song sheets/books. The typical reader might have come into possession of existing ChordPro songs they want to use, or previously laid out songs/books using wordprocessing software - a useful benchmark for addressing and reframing functional/formatting expectations might be "But I can do that easily in MS-Word!"

**Scope:** Aims to introduce all the core ChordPro areas (GUI, song, configuration, CLI) and terminology, but not cover each one in full depth, which the following sections will cover. Primarily focused on GUI usage, taking the learning journey as far as possible within the (new) GUI. Brief (advanced) introduction to ChordPro CLI for users already familiar with using terminal/CLI, but not a command line tutorial for all different platforms. It should equip users with a good _conceptual_ foundation, so they can make effective, practical use of Cookbook and Reference.

**Organisation:** Linear structure that leads users through a clear and logical learning journey, in which each new idea or technique builds on those introduced previously.

**Method:** Narrative explanation in clear, informal language, supported by illustration/screenshots and downloadable sample songs/configs. Screenshots taken from Windows (as likely largest potential user base?), but platform-neutral (e.g. cropped to content, without window frame/buttons.)

**Workplan:** I (Alyn) will take on the whole job, as it makes sense to have a single 'guiding voice' through the learning process. I'll share my preparatory thinking in the form of topic outlines, and welcome feedback on the learning journey these represent, which is an important aspect of the Tutorial. Remember, this is for non-technical users so please review and comment from the perspective of someone who's rarely used a text editor and almost never the command line, e.g. if you know musicians who like the idea of ChordPro, but can't get to grips with it in practice.

It's possible that process of "walking in the novice's shoes" might also throw up potential enhancements, to feed into GUI development.

## 2. Cookbook

**Rationale:** Anticipate practical questions that commonly arise, e.g. in the User Forum. Essentially an expansion of the existing Hints and Tips section. Many forum questions ask "How do I...?", often focused on achieving particular output results. Beyond the core knowledge (Tutorial), users will have different priorities for how they want to use ChordPro, so this approach seems a useful way to serve a variety of user needs. This section might get used often at basic/intermediate level, and maybe less often by experienced users (either they know it, or have customised their song production as far as they need).

**Focus:** Task-centred, *i.e.* how to achieve specific results with ChordPro.

**Target audience:** Intermediate users, familar with basic ChordPro concepts and practices (maybe CLI?), and have formed basic design/layout idea from using wordprocessor/layout software (though Tutorial should have taught them that ChordPro is a different category of software).

**Scope:** Potentially unlimited, but if it gets too big, maybe limit to the tasks/questions the recur most often, so the user forum can focus on rare or one-off cases.

**Organisation:** Non-linear, with specific, standalone tasks (one per page?), maybe linked from a 'Contents' list. Assumed prior knowledge is what's covered in the Tutorial (links to other recipes or Reference optional but not required). As examples grow, it might become clearer how they can be grouped under specific headings (e.g. Page Layout, Header/Footer content, Songbooks, Exporting or Other Output formats etc.). Each topic includes links to relevant sections in Reference (enabling readers to customise the example used).

**Method:** One or more ways to achieve the given outcome/task (e.g. song-based and config-based, if applicable), including copyable code snippets or illustrations, ideally supported by explanation of how/why it works. The solution might involve a combination of song source and configuration (and sometimes CLI) techniques, so this section shows how to combine those broad areas of knowledge in useful ways. Or, some outcomes could be achieved within a song _or_ through configuration, so we provide 2+ ways to do the task (maybe explaining pros and cons of each).

**Workplan:** This will follow from the Tutorial, giving the main knowledge foundation. I'll provide a general format and principles for recipes, but anyone can contribute specific tasks/tricks that they've found useful or address common questions. I'll take editorial responsibility, to ensure some coherence across the whole Cookbook (and therefore clarity for readers/users).

## 3. Reference

**Rationale:** The 'gold standard' for complete and accurate documenting of each aspect of ChordPro directives, configuration options etc. This section might get used frequently by experienced users, and rarely by beginnenrs.

**Focus:** Software-centred, *i.e.* how ChordPro works, at the most fine-grained level. Accuracy is key.

**Target audience:** Experienced users familiar with the main ChordPro concepts (Turorial) and some practical usage (+Cookbook), who need to check syntax or detailed functionality.

**Scope:** Everything that can be done with ChordPro, assuming the reader has a good grasp of ChordPro concepts and functionality.

**Organisation:** As currently, i.e. reflecting different areas of functionality, e.g. Chords, Directives, Configuration, CLI options.

**Method:** Detailed description of each aspect of ChordPro functionality.

**Workplan:** Continue with current approach for now, as that's detailed and accurate, and in sync with software development. Maybe a little light restructuring, as a few existing topics might get replaced by the above two sections.

## Other topic areas

Some important material might not fit neatly into the above structure. These gaps will become clearer as other sections develop, so will probably be part of a final tidy-up.

* FAQ: any other general questions, which aren't already covered by the "How do I ...?" in the Cookbook
* Further questions -> User Forum; Bug reports -> GitHub
* Useful links
* Future developments

Out of scope
------------

* Website 'shop window' that promotes *what* ChordPro can do, so documentation focuses on *how or why to do it* (but might take website as a benchmark, so users can actually achieve the things they've been told ChordPro can do).
* How to use terminal and command line _in general_. CLI guidance limited to using ChordPro on command line. (In future, possibility of built-in terminal window/pane might help make this more accessible).
* Translation / GUI localisation. These might be useful ideals at some future point, but probably the immediate priority is for English-language versions to develop and stabilise.

# Process

Production methods
------------------
For compatibility with existing documentation/website based  on Hugo processing of Markdown:

* Markdown for structured text
* Mermaid for diagrams (?)
* SVG for vector illustrations
* PNG (200dpi?) for raster images/screenshots

Any other thoughts/technologies for process and production?

**Comment by gwyndaf** [comment]
@sciurius I'm approaching a point in writing the Tutorial where it seems sensible to share some material for review:

1. By developers and experienced users, to verify conceptual and practical accuracy.
2. By non-technical users, to verify it's clear and easy to understand (both concepts and practices of using ChordPro).

2, in particular, favours having the documents accessible via the web (beta site only at this draft stage). I don't expect non-technical users to be familiar with using GitHub or reading/interpreting Markdown documents. But a key purpose is to help less-technical users start using ChordPro. I know a few, who don't currently use ChordPro or engage in its existing discussions, so I'll need to send them a link for them to review it online.

So...

**Overall process**
Is it feasible to have new documentation available at the beta site, but independent of updates to existing documentation?

I'm guessing beta site content gets pulled into the main site with each release cycle, but the timing of my work won't necessarily align with that. So Tutorial etc. needs to remain on the beta site as they evolve, even as other site content is migrated to the release site. Is that possible?

**Git mechanics and protocol**
Currently, I'm working with a local git repository and a hugo setup that includes the ChordPro theme, so I can preview the presentation, as well as content, as I create and edit the material.

But, I'm not a developer, so don't know the process and protocol for merging those changes into the existing ChordPro respository. My existing git knowledge is largely one-way (i.e. `pull`).

I'm guessing I might need to create a `usersguide` branch in my pulled copy of the GitHub repository, and put my files there.

But, I don't know if I can just `push` those changes to the GitHub repository. I understand there's a 'pull request' protocol, but don't really know how to use it. Can you clarify your preferred process?

An alternative is that I have my own GitHub repository for the User's Guide, from which you can pull material into the main repository. Is that more sensible?

**Likely changes**
I'm planning a separate `usersguide` subfolder, so the User's Guide pages will be self-contained and neatly organised, separate from existing documentation files.

However, I have made some small CSS additions, to support presentation methods that I need, but aren't in the current site-wide CSS. These include deeper heading levels, and styling `blockquote` elements for tips or incidental notes (less prominent than main content - effectively editorial 'sidebars', except inline, not at the side).

These don't change existing definitions in `assets/css/site.css`, and simply add new ones. But, is it preferable to create separate CSS for the User's Guide? That'll need a bit more Hugo learning by me (and refreshing rusty CSS knowledge), but I imagine is feasible.

**Comment by sciurius** [comment]
I'll get back to this asap.

One thing: Our current Hugo structure is based on a ancient approach, now widely superceded by Hugo themes. A while ago I investigated some alternatives and found the [Relearn](https://mcshelby.github.io/hugo-theme-relearn/index.html) theme a good candidate. I've postponed migrating the web site + current docs awaiting your progress and feedback. Can you take a look at that theme to see if it fits your requirements?

**Comment by sciurius** [comment]
> Git mechanics and protocol

At the moment our web site looks more or less like
````
index.html   -- landing page, copy of chordpro/landing.html, links to chordpro/index.html
chordpro/    ...all the docs for the current version...
beta/        ...all the docs for the beta version...
````

Let's keep it as a separate repo, e.g. `chordpro-userguide` and publish it to a side directory:
````
index.html   -- landing page, copy of chordpro/landing.html, links to chordpro/index.html
chordpro/    ...all the docs for the current version...
chordpro-userguide/    ...
beta/        ...all the docs for the beta version...
beta-userguide/        ...
````

That will be easier to get started with, since the two repos are updated independently and you don't have to watch for changes.

**Comment by ChordPro** [comment]
I've sent you an ivite for the chordpro/userguide repo. Instructions follow.

**Comment by ChordPro** [comment]
According to github, this should work:
````
git remote add origin https://github.com/ChordPro/userguide.git
git branch -M main
git push -u origin main
````

**Comment by gwyndaf** [comment]
Fantastic, thanks Johan. I can deal with pushing on a solo basis, just not familiar with mechanisms for working collaboratively, so this seems simplest/safest :)

Fortunately, my Hugo learning is since it used themes, so I've got the existing style as a ChordPro theme, and it'll be easy to add and review the Relearn one.

**Comment by gwyndaf** [comment]
> Our current Hugo structure is based on a ancient approach, now widely superceded by Hugo themes

So, it sounds like a _functional_ reason for updating, yes? That makes sense. I can see that some features are supported out-of-the-box, for which I needed to tweak the Hugo config or `site.css` for the existing style.

_Stylistically_, the Relearn theme seems better suited to 'brochure' usage than documentation. It looks very impressive and elegant (except for italics), with lots of space, _but_ very inefficient for communication. From a very rough test, I get ~25% less content on a single (desktop) screen, compared with the existing style.

That means a user needs to do more scrolling and it's harder to gain a quick overview, even for a relatively short topic. So, subjectively, it feels harder to read and assimilate information, especially if I need to explain new concepts for non-technical users. That could create users the impression that ChordPro is hard to use, rather than helping users understand it.

But, I'm still getting used to Hugo and its themes, so not totally sure what classes as 'part of Hugo' or 'core to the theme' or 'in the theme, but easily configurable'. It might be that the stylistic aspects can easily be adjusted, while still benefiting from new functionality.

**Comment by sciurius** [comment]
Feel free to choose any other hugo style if that fits better.

**Comment by gwyndaf** [comment]
OK. Given that my understanding of Hugo only scratches the surface, that might involve a separate chunk of work.

I can meet immediate communication needs with a few small additions I've made to the existing `hugo.toml` and `site.css` (but might investigate how I can make Tutorial-specific CSS for those tweaks).

Unless there's a pressing need to update the site style, my feeling is the following workplan to complete the Tutorial:

* a1: Quick Start (GUI program - install and operate)
* a2: ChordPro basics (concepts and song _description_ format)
* a4: Lead sheet _formatting_ (in-song directives/markup) - independent of GUI
* a3: Using the program (GUI) - including current refinements
* b: Customisation (configuration) concepts and selected techniques
* **Review Hugo/theme**
* c: Advanced (deeper dives into some earlier topics, as a foundation for using the Reference)

Does that seem a sensible approach?

**Comment by sciurius** [comment]
Fortunately, the Hugo markdown sources are mostly (if not completely) independent of the theme used. So the theme review can be done at almost any time.

IIRC there is one important shortcode that must be added anyway: `showpage` (or `showpages`) that shows a small version of a (larger) image to be expanded with a mouse click. You'll like it.

**Comment by gwyndaf** [comment]
Yes, most of my content is near-pure markdown. That's probably one reason I've not explored Hugo's greater depths yet.

My main addition was to copy the metadata format of existing docs (between `---` pairs), but I got some strange effects with the Relearn theme, and it seems that `+++` is now preferred framing.

So, I'll probably focus on the content (with simple markdown) for now, and then refine metadata and navigation/interaction touches (`showpages` noted).
